server/src/main/java/org/candlepin/model/EntitlementCurator.java

```diff
@@ -437,27 +437,61 @@ private String sanitizeMatchesFilter(String matches) {
             List<String> entitlementIds = criteria.list();
 
             if (entitlementIds != null && !entitlementIds.isEmpty()) {
-                criteria = this.currentSession()
-                    .createCriteria(Entitlement.class)
-                    .add(CPRestrictions.in("id", entitlementIds));
+                entitlementsPage = createPage(this.findEntitlements(entitlementIds), pageRequest);
 
-                entitlementsPage = listByCriteria(criteria, pageRequest);
             }
             else {
                 entitlementsPage = new Page<>();
-                entitlementsPage.setPageData(Collections.<Entitlement>emptyList());
+                entitlementsPage.setPageData(Collections.emptyList());
                 entitlementsPage.setMaxRecords(0);
             }
         }
-
         return entitlementsPage;
     }
 
+    private List<Entitlement> findEntitlements(Iterable<String> entitlementIds) {
+        String queryStr = "SELECT e " +
+            "FROM Entitlement e " +
+            "WHERE e.id IN (:entitlement_ids) " +
+            "ORDER BY e.id";
+
+        Query query = getEntityManager().createQuery(queryStr, Entitlement.class);
+        List<Entitlement> result = new LinkedList<>();
+        if (entitlementIds == null || !entitlementIds.iterator().hasNext()) {
+            return result;
+        }
+
+        int blockSize = Math.min(this.getInBlockSize(), this.getQueryParameterLimit() / 2);
+        for (List<String> block : Iterables.partition(entitlementIds, blockSize)) {
+            result.addAll(query.setParameter("entitlement_ids", block).getResultList());
+        }
+        return result;
+    }
+
+    private Page<List<Entitlement>> createPage(List<Entitlement> entitlements, PageRequest pageRequest) {
+        Page<List<Entitlement>> page = new Page<>();
+        if (pageRequest != null) {
+            int start = (pageRequest.getPage() - 1) * pageRequest.getPerPage();
+            List<Entitlement> currentPage = entitlements.stream()
```

Comment:
If we're requiring a list as input, we could use .sublist instead of streaming and collecting.